### ISSUE TITLE
Undefined behaviour sanitiser issues

### ISSUE THREAD

**Issue by astamm** [astamm/fdacluster#2]
Prof. Ripley detected these issues: https://cran.r-project.org/web/checks/check_results_fdacluster.html.

I am investigating any coding error in `fdacluster` but I wonder whether it might be an issues in submatrix views in `armadillo`. Do you know of such issues @eddelbuettel ?

**Comment by eddelbuettel** [comment]
No I do not.  That said, as @conradsnicta updates at a cadence faster than CRAN's desired monthly (the do check for 'no more than six uploads in six months' so I space it out) we are one bug fix behind and/or it could be a new issue.

So ideally could you please
- replicate BDR's findings using a properly instrumented Docker container, the ones by Winston in https://github.com/wch/r-debug are good (I also produce(d) some but not as comprehensive or regularly as he does now so I defer ;-) )
- see if RcppArmadillo 0.10.8.2.0.0 in branch / master (which I'll push to master in a few minutes, full rev.dep check finished overnight) does still show it
- maybe even distill it to a plain C++ example for @conradsnicta ?

I can help with some of the Docker mechanics if you need it.

**Comment by astamm** [comment]
I am happy to implement those steps although, as you might have hinted by offering your help, I am not that familiar with Docker mechanics. I am happy to accept your help here. I will still read through Winston repo, thanks.

**Comment by eddelbuettel** [comment]
Can you get access to a Linux machine that has Docker enabled?   I understand this foreign territory I never visit called 'macOS' also does but I can't help with specific there.

Docker in general will use something like the following to _make your current working directory_ (ie your repo) available to it:

     docker run --rm -ti -v ${PWD}:/mywork -w /mywork --security-opt seccomp=unconfined wch1/r-debug bash

and then (important !!) use the `RDsan` or `RDcsan` versions of R instrumented for it.  _You then need to install all your depenencies_ using that instance.  As I recall, the container is 'fat' and has a few commonly used things already (such as `Rcpp`) but you likely need `RcppArmadillo`, `nloptr`, ...  so it is a good idea to keep the Docker session up in a terminal while you iterate over the code in the directory you mounted.  That way you have toolchain to edit outside yet Docker giving you (UB)SAN equipped R.

**Comment by astamm** [comment]
Thanks Dirk. I do not have access to a Linux machine. I am installing Docker and my Mac and I hope that between Winston's `README` and your helpful comments, I will be able to get it running.

**Comment by eddelbuettel** [comment]
Yes. There are approximately 1e4 tutorials for Docker use on macOS.  I understand there is a GUI.  I only ever drive from the cmdline. 

Also, on the topic, a classic from 1995 for you here:  
![image](https://user-images.githubusercontent.com/673121/152170954-7845cdf3-741f-4ce5-87c2-8c6ea6b38c80.png)

**Comment by astamm** [comment]
That's a good one!
I have indeed downloaded the GUI. And I ran `docker run --rm -ti --security-opt seccomp=unconfined wch1/r-debug` in my repo from terminal as per Winston `README` Quickstart instruction. It should then pop up in the GUI, so they say in there. But it is taking quite some time downloading lots of stuff apparently. Should take some time.

**Comment by eddelbuettel** [comment]
Docker is much 'skinnier' and faster on Linux and it is Linux-on-Linux there.  You get the pleasure of taking it all in via a VM slipped in between Docker and your OS.

The PR I mentioned is now up at https://github.com/RcppCore/RcppArmadillo/pull/363.  Those were commits I had made here once 10.8.2 came out but not pushed til the (length, slow machine) reverse-depends checks were done.   Congrats, btw, on another package at CRAN :)  I saw this via [CRANberries](https://github.com/RcppCore/RcppArmadillo/pull/363) (via RSS and Twitter).

**Comment by astamm** [comment]
Okay, the Docker image is mounted. I can see my repo in there but `RDsan` fails because of the emulation... I will have to dig into the online tutorials to understand how to move forward.

**Comment by eddelbuettel** [comment]
Ah, could be the VM gets in the way.  RHub has a sanitizer build available but it may not replicate what BDR found.  Worth a try.

**Comment by conradsnicta** [comment]
@astamm @eddelbuettel Can you provide a small program in pure C++ that can replicate this issue (ie. without use of R and Rcpp) ?  Please file a bug report at the Armadillo repo: https://gitlab.com/conradsnicta/armadillo-code

I don't really have the time to dig through your code, but I had a brief look at the error log ( https://www.stats.ox.ac.uk/pub/bdr/memtests/clang-UBSAN/fdacluster/tests/testthat.Rout ) and this pops out:

> `/data/gannet/ripley/R/test-clang/RcppArmadillo/include/armadillo_bits/Mat_meat.hpp:6076:22: runtime error: reference binding to null pointer of type 'const double'`

Without looking at your code, the above suggests that a matrix or vector element is being accessed in a zero-sized (empty) matrix/vector  Maybe a matrix/vector wasn't correctly initialised?  In addition, is the element access being done via `X[index]` (note the square brackets), or perhaps `X.at(row,col)` ?

It's also possible that this is a false positive, but it's still prudent to thoroughly check this.

**Comment by conradsnicta** [comment]
@astamm update: I believe the bug is your code. I did a very quick source code lookup from the error log mentioned previously, which leads to this code:

https://github.com/astamm/fdacluster/blob/e3861b593a9ba19465de183c8ac2372c1f45d3b8/src/baseOptimizerClass.cpp#L91-L97

UBSAN complains that line 94 is a problem: `&(initialParameters[0])`.  The `initialParameters` variable is defined as `arma::rowvec&` at the entry to the function.  Note the `if` statement enclosing this code: `if (initialParameters.size() == 0)`.

UBSAN seems to be right here. Since `initialParameters` is empty (has a size of zero), it has no elements.  So accessing the element at index 0 via `initialParameters[0]` doesn't make sense, as there is no element at index 0.  In other words, the problematic statement `&(initialParameters[0])` obtains an invalid memory address.

It seems this problematic pattern is repeated elsewhere in your codebase.  See for example: https://github.com/astamm/fdacluster/blob/e3861b593a9ba19465de183c8ac2372c1f45d3b8/src/baseOptimizerClass.cpp#L143-L149

Note that accessing elements in Armadillo vectors and matrices via `[index]` (ie. square brackets) deliberately omits bounds checks.  I suggest using `(index)` instead (ie. round brackets), which has a run-time bounds check.  http://arma.sourceforge.net/docs.html#element_access

**Comment by astamm** [comment]
Thanks for digging into it @conradsnicta. I was doing the same line of reasoning here. I will check whether the other runtime errors spotted by UBSAN are only corollaries of that one or if there are other issues in my code.

**Comment by astamm** [comment]
I usually use round brackets as per your recommandations in armadillo documentation indeed. That one slipped away.

**Comment by astamm** [comment]
Apparently, I did not fix all of them.
I need to find a Linux machine to run the docker image to check these as BDR did.

**Comment by astamm** [comment]
I have access to a Linux computer with `singularity` but not `docker`. @eddelbuettel Do you know if that would work for running Winston's r-devel image?
I tried r-hub's but their Debian machine which is used for compiling with sanitizers has trouble installing `Rcpp` and `RcppArmadillo` apparently. Gabor is looking into it.

**Comment by eddelbuettel** [comment]
Sorry, honestly no idea. I often hear from `singularity` users who proseculize on how much better yada yada yada. I mostly tune out by now as I use Docker which "works for me (TM)".  I think it "should" work for you but then we said the same for macOS where it turned out the virtualization got in the way of ASAN/UBSAN.

I use that r-hub machine fairly regularly, and I think I only ever had issues with missing external libraries. There is something persistently "off" though (but in a benign way) -- I think each email I get from it says 'Error' but when you look at the log it is the inverse.  Free service and all that.   @gaborcsardi will likely help you; the bigger issue is that CRAN is not exactly as helpful as one would hope and that box is not a 100% replica of what BDR uses.  Ah well.
